source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -20,102 +20,132 @@
+    void (*heap_failure_callback)(heap_fail_t);
+    NS_LIST_HEAD(hole_t, link) holes_list;
+    ns_mem_heap_size_t heap_size;
+} ns_mem_book_t;
+
+static ns_mem_book_t *static_context; // heap pointer for original "ns_" API use
 
 // size of a hole_t in our word units
-#define HOLE_T_SIZE ((sizeof(hole_t) + sizeof(int) - 1) / sizeof(int))
+#define HOLE_T_SIZE ((sizeof(hole_t) + sizeof(ns_mem_word_size_t) - 1) / sizeof(ns_mem_word_size_t))
 
-static NS_INLINE hole_t *hole_from_block_start(int *start)
+static NS_INLINE hole_t *hole_from_block_start(ns_mem_word_size_t *start)
 {
     return (hole_t *)(start + 1);
 }
 
-static NS_INLINE int *block_start_from_hole(hole_t *start)
+static NS_INLINE ns_mem_word_size_t *block_start_from_hole(hole_t *start)
 {
-    return ((int *)start) - 1;
+    return ((ns_mem_word_size_t *)start) - 1;
 }
 
-
-static void heap_failure(heap_fail_t reason)
+static void heap_failure(void (*callback)(heap_fail_t), heap_fail_t reason)
 {
-    if (heap_failure_callback) {
-        heap_failure_callback(reason);
+    if (callback) {
+        callback(reason);
     }
 }
 
 #endif
 
-void ns_dyn_mem_init(uint8_t *heap, uint16_t h_size, void (*passed_fptr)(heap_fail_t), mem_stat_t *info_ptr)
+void ns_dyn_mem_init(uint8_t *heap, uint16_t h_size,
+                     void (*passed_fptr)(heap_fail_t), mem_stat_t *info_ptr)
+{
+    static_context = ns_mem_init((void *)heap, h_size,
+                                      (void (*)(heap_fail_t))passed_fptr,
+                                                   (mem_stat_t *)info_ptr);
+}
+
+const mem_stat_t *ns_dyn_mem_get_mem_stat(void)
+{
+#ifndef STANDARD_MALLOC
+    return (mem_stat_t *)ns_mem_get_mem_stat(static_context);
+#else
+    return NULL;
+#endif
+}
+
+
+ns_mem_book_t *ns_mem_init(void *heap, ns_mem_heap_size_t h_size,
+                         void (*passed_fptr)(heap_fail_t),
+                                      mem_stat_t *info_ptr)
 {
 #ifndef STANDARD_MALLOC
-    int *ptr;
-    int temp_int;
+    ns_mem_book_t *book;
+
+    ns_mem_word_size_t *ptr;
+    ns_mem_word_size_t temp_int;
     /* Do memory alignment */
-    temp_int = ((uintptr_t)heap % sizeof(int));
+    temp_int = ((uintptr_t)heap % sizeof(ns_mem_word_size_t));
     if (temp_int) {
-        heap += (sizeof(int) - temp_int);
-        h_size -= (sizeof(int) - temp_int);
+        heap += (sizeof(ns_mem_word_size_t) - temp_int);
+        h_size -= (sizeof(ns_mem_word_size_t) - temp_int);
     }
 
     /* Make correction for total length also */
-    temp_int = (h_size % sizeof(int));
+    temp_int = (h_size % sizeof(ns_mem_word_size_t));
     if (temp_int) {
-        h_size -= (sizeof(int) - temp_int);
+        h_size -= (sizeof(ns_mem_word_size_t) - temp_int);
     }
-    heap_main = (int *)heap; // SET Heap Pointer
-    heap_size = h_size; //Set Heap Size
-    temp_int = (h_size / sizeof(int));
+    book = (ns_mem_book_t *)heap;
+    book->holes_list.slist.first_entry = NULL;
```

Comment:
Um, no. This is ns_list_init hard-coded, which you've already got below. I guess the original code both statically initialised it with NS_LIST_DEFINE, then dynamically with the ns_list_init, which is how you've ended up doing it twice.